File: Desafios/Logica/Fibonacci/R/fibonacci.R
fibonacci <- local({
    # Ambiente para armazenar valores já calculados
    cache <- new.env()
    
    function(number) {
        if (exists(as.character(number), envir = cache)) {
            # Retorna o valor do cache se já calculado
            return(get(as.character(number), envir = cache))
        }
        
        # Calcular o valor de Fibonacci
        result <- if (number == 0 | number == 1) {
            number
        } else {
            fibonacci(number - 1) + fibonacci(number - 2)
        }
        
        # Armaneza o valor no cache
        assign(as.character(number), result, envir = cache)
        return(result)
    }
})


# Solicitar um número ao usuário
cat("Digite um número: ")
n <- as.integer(readline())

# Calcular e exibir o resultado
cat(paste("O número Fibonacci na posição", n, "é:", fibonacci(n), "\n"))

File: Desafios/Logica/Fibonacci/C/fibonacci.c
#include <stdio.h>
#include <locale.h>

int fibonacci(int number);

int main(int argc, char *argv[]) {
	
	int n;
	
	setlocale(LC_ALL, "Portuguese");
	printf("Digite um número: ");
	scanf("%d", &n);
	
    printf("\nO número Fibonacci na posição %d é: %d\n\n", n, fibonacci(n));
	system("pause");
	
	return 0;

}

int fibonacci(int number) {
    if (number == 0 || number == 1) {
               return number;
    } else {
           return fibonacci(number - 1) + fibonacci(number - 2);
    }
}
